### PR TITLE
Complement ATen operators' variants with `func`, `inplace` and `out`.

### PR DESCRIPTION
**Pull request by xytintel**
1. Complement ATen operators' variants implementation.
2. Add new Binary operators, e.g. rsub, remainder, fmod, reciprocal, pow.
dep on: https://github.com/intel/torch-xpu-ops/pull/42

### CODE REVIEW

File: src/aten/sycl/BinaryKernels.cpp [hunk 1]
```diff
@@ -42,7 +42,43 @@ struct DivFunctor {
...
+    if (r != 0 && c10::signs_differ(r, b)) {
+      r += b;
+    }
+    return r;
+  }
+};
+
+template <typename scalar_t>
+struct RemainderFloatingFunctor {
+  scalar_t operator()(scalar_t a, scalar_t b) const {
```
Comment (fengyuan14) [inline src/aten/sycl/BinaryKernels.cpp:58]: Please check `__ubsan_ignore_float_divide_by_zero__`, which is used by both cpu and cuda kernels.

File: src/aten/sycl/BinaryKernels.cpp [hunk 2]
```diff
@@ -42,7 +42,43 @@ struct DivFunctor {
   }
 };
 
-void add_kernel(TensorIterator& iter, const c10::Scalar& alpha) {
+template <typename scalar_t>
```
Comment (fengyuan14) [inline src/aten/sycl/BinaryKernels.cpp:45]: Please align CUDA, let us put Binary remainder kernel into `BinaryRemainderKernel.cpp`, like `aten/src/ATen/native/cuda/BinaryRemainderKernel.cu`.

File: src/aten/sycl/BinaryKernels.cpp [hunk 3]
```diff
@@ -42,7 +42,43 @@ struct DivFunctor {
...
+struct FmodIntegralFunctor {
+  scalar_t operator()(scalar_t a, scalar_t b) const {
+    return a % b;
+  }
+};
+
+template <typename scalar_t>
+struct FmodFloatingFunctor {
+  scalar_t operator()(scalar_t a, scalar_t b) const {
```
Comment (fengyuan14) [inline src/aten/sycl/BinaryKernels.cpp:76]: `__ubsan_ignore_float_divide_by_zero__`

File: src/aten/Pow.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  auto iter = TensorIterator::binary_op(result, base, exp);
+  native::xpu::pow_tensor_tensor_kernel(iter);
+  return result;
+}
+
+Tensor& XPUNativeFunctions::pow_out(
+    const Scalar& base,
+    const Tensor& exp,
+    Tensor& result) {
+  if (base.isComplex() && base.toComplexDouble() == 1.0) {
```
Comment (fengyuan14) [inline src/aten/Pow.cpp:24]: Please follow the helpers usage.
image

File: src/aten/Pow.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        native::wrapped_scalar_tensor(base), exp, result);
+  }
+  return result;
+}
+
+Tensor& XPUNativeFunctions::pow_out(
+    const Tensor& self,
+    const Scalar& exponent,
+    Tensor& result) {
+  auto iter = TensorIterator::unary_op(result, self);
```
Comment (fengyuan14) [inline src/aten/Pow.cpp:39]: Please align with the helpers usage,
image

File: src/aten/sycl/BinaryKernels.cpp [hunk 6]
```diff
@@ -59,11 +95,11 @@ void add_kernel(TensorIterator& iter, const c10::Scalar& alpha) {
   }
 }
 
-void sub_kernel(TensorIterator& iter, const c10::Scalar& alpha) {
+void sub_kernel(TensorIteratorBase& iter, const c10::Scalar& alpha) {
```
Comment (fengyuan14) [inline src/aten/sycl/BinaryKernels.cpp:98]: Why using TIbase? not TI? CUDA is using it?

File: src/aten/sycl/PowKernels.cpp [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+      result *= a;
+    }
+    b /= 2;
+    a *= a;
+  }
+  return result;
+}
+
+template <typename scalar_t>
+inline scalar_t pow_impl(scalar_t a, scalar_t b) {
```
Comment (fengyuan14) [inline src/aten/sycl/PowKernels.cpp:30]: We need align scalar behavior with CUDA.
image

File: src/aten/sycl/PowKernels.cpp [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+
+template <typename scalar_t>
+struct PowImplUnaryFunctor3 {
+  scalar_t operator()(scalar_t base) const {
+    return 1.0 / (base * base);
+  }
+};
+
+template <typename scalar_t>
+struct PowImplUnaryFunctor4 {
```
Comment (fengyuan14) [inline src/aten/sycl/PowKernels.cpp:202]: Remove if no used.

File: src/aten/sycl/PowKernels.cpp [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+  // .5 (sqrt), -.5 (rsqrt) and -1 (reciprocal) specializations are handled
+  // in pow_tensor_scalar_kernel
+  if (d_exp == 2) {
+    gpu_kernel(iter, PowImplUnaryFunctor1<Base_type>());
+  } else if (d_exp == 3) {
+    gpu_kernel(iter, PowImplUnaryFunctor2<Base_type>());
+  } else if (d_exp == -2) {
+    gpu_kernel(iter, PowImplUnaryFunctor3<Base_type>());
+  } else {
+    auto f = PowScalarTensorFunctor2<Base_type>(exp);
```
Comment (fengyuan14) [inline src/aten/sycl/PowKernels.cpp:231]: CUDA is using `pow_` for all default `exp` and `dtype` path? We need align with CUDA.

File: src/aten/sycl/PowKernels.cpp [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+    }
+  }
+  if (isComplexType(iter.common_dtype()) || exp_scalar.isComplex()) {
+    if (iter.common_dtype() == kComplexHalf) {
+      pow_chalf_tensor_scalar_impl(iter, exp_scalar);
+      return;
+    }
+    AT_DISPATCH_COMPLEX_TYPES(iter.common_dtype(), "pow_xpu", [&]() {
+      const auto exp = exp_scalar.to<scalar_t>();
+      gpu_kernel(iter, PowScalarTensorFunctor2<scalar_t>(exp));
```
Comment (fengyuan14) [inline src/aten/sycl/PowKernels.cpp:256]: `pow_`

File: src/aten/sycl/UnaryKernels.cpp [hunk 11]
```diff
@@ -18,6 +18,112 @@ struct AbsFunctor {
...
+};
+
+template <typename scalar_t>
+struct NegFunctor {
+  scalar_t operator()(scalar_t a) const {
+    return -a;
+  }
+};
+
+template <typename scalar_t>
```
Comment (fengyuan14) [inline src/aten/sycl/UnaryKernels.cpp:87]: Comment:
// For these common numeric helpers, we should share with CUDA code when putting XPU in-tree. `aten/src/ATen/native/cuda/UnaryFractionKernels.cu :69`

File: src/aten/sycl/UnaryKernels.cpp [hunk 12]
```diff
@@ -36,4 +142,131 @@ void abs_kernel(TensorIteratorBase& iter) {
   }
 }
 
+void sin_kernel(TensorIteratorBase& iter) {
+  auto common_dtype = iter.common_dtype();
+  if (at::isComplexType(common_dtype)) {
+    AT_DISPATCH_COMPLEX_TYPES_AND(
+        kComplexHalf, common_dtype, "sin_name", [&]() {
```
Comment (fengyuan14) [inline src/aten/sycl/UnaryKernels.cpp:149]: `sin_name` -> `sin_xpu`

File: src/aten/sycl/UnaryKernels.cpp [hunk 13]
```diff
@@ -36,4 +142,131 @@ void abs_kernel(TensorIteratorBase& iter) {
...
+          gpu_kernel(iter, SinFunctor<scalar_t>());
+        });
+  }
+}
+
+void cos_kernel(TensorIteratorBase& iter) {
+  auto common_dtype = iter.common_dtype();
+  if (at::isComplexType(common_dtype)) {
+    AT_DISPATCH_COMPLEX_TYPES_AND(
+        kComplexHalf, common_dtype, "cos_name", [&]() {
```
Comment (fengyuan14) [inline src/aten/sycl/UnaryKernels.cpp:165]: `cos_xpu`

File: src/aten/sycl/UnaryKernels.cpp [hunk 14]
```diff
@@ -36,4 +142,131 @@ void abs_kernel(TensorIteratorBase& iter) {
...
+        "rsqrt_xpu",
+        [&]() { gpu_kernel(iter, RsqrtFunctor<scalar_t>()); });
+  }
+}
+
+void tanh_kernel(TensorIteratorBase& iter) {
+  auto common_dtype = iter.common_dtype();
+  if (at::isComplexType(common_dtype)) {
+    AT_DISPATCH_COMPLEX_TYPES_AND(
+        kComplexHalf, common_dtype, "tanh_name", [&]() {
```
Comment (fengyuan14) [inline src/aten/sycl/UnaryKernels.cpp:235]: `tanh_xpu`

Comment (fengyuan14) [inline src/aten/sycl/UnaryKernels.cpp:235]:
```suggestion
        kComplexHalf, common_dtype, "tanh_xpu", [&]() {
```

File: src/aten/sycl/UnaryKernels.cpp (hunk 13 above)
Comment (fengyuan14) [inline src/aten/sycl/UnaryKernels.cpp:165]:
```suggestion
        kComplexHalf, common_dtype, "cos_xpu", [&]() {
```

File: src/aten/sycl/UnaryKernels.cpp (hunk 12 above)
Comment (fengyuan14) [inline src/aten/sycl/UnaryKernels.cpp:149]:
```suggestion
        kComplexHalf, common_dtype, "sin_xpu", [&]() {
```

File: src/aten/sycl/BinaryKernels.cpp (hunk 6 above)
Comment (xytintel) [reply src/aten/sycl/BinaryKernels.cpp:98]: Align to cuda's impl

File: src/aten/sycl/PowKernels.cpp [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+      const auto base = iter.scalar_value<scalar_t>(1);
+      iter.remove_operand(1);
+      pow_scalar_tensor_impl(iter, base);
+    } else if (iter.is_cpu_scalar(2)) {
+      const auto exp = iter.scalar_value<scalar_t>(2);
+      iter.remove_operand(2);
+      pow_chalf_tensor_scalar_impl(iter, exp);
+    } else {
+      TORCH_INTERNAL_ASSERT(!iter.is_cpu_scalar(1) && !iter.is_cpu_scalar(2));
+      auto f = PowTensorTensorCastFunctor<scalar_t>();
```
Comment (fengyuan14) [inline src/aten/sycl/PowKernels.cpp:131]: opmath.

Comment (xytintel) [reply src/aten/sycl/PowKernels.cpp:131]: Align with CUDA in kernel

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**